app/javascript/serviceworker/simple-crypto.ts

```diff
@@ -0,0 +1,119 @@
+ *
+ * Usage:
+ *     const secret = new SimpleCrypto("my-passphrase", "my-salt");
+ *     await secret.init();
+ *     const encrypted = await secret.encrypt("hello world");
+ *     const decrypted = await secret.decrypt(encrypted);
+ */
+export class SimpleCrypto {
+  private passphrase: string;
+  private salt: string;
+  private key?: CryptoKey;
+
+  /**
+   * @param passphrase The passphrase to use. This can be provided by the user and should be stored securely.
+   * @param salt A preferably random string that is used to salt the passphrase, to ensure that ciphertexts of the same data with the same passphrase are different.
+   * @returns A new SimpleCrypto instance.
+   */
+  constructor(passphrase: string, salt: string) {
+    this.passphrase = passphrase;
+    this.salt = salt;
+  }
+
+  /**
+   * Parses the passphrase and salt into a determinstic CryptoKey that can be used for encryption and decryption.
+   *
+   * @returns A Promise that resolves when the key is ready.
+   * */
+  public async init(): Promise<void> {
+    const encoder = new TextEncoder();
+
+    // First, import an initial key from the passphrase. On its own, this key
+    // is not secure, as it could be brute forced, so we can't directly use it
+    // to encrypt and decrypt data.
+    const initialKey = await crypto.subtle.importKey(
+      "raw",
+      encoder.encode(this.passphrase),
+      { name: "PBKDF2" },
+      false,
+      ["deriveBits", "deriveKey"]
+    );
+
+    // Derive a new key from the initial key. Using salt makes it unique per
+    // user, and using lots of iterations makes it computationally expensive to
+    // brute force.
+    this.key = await crypto.subtle.deriveKey(
+      {
+        name: "PBKDF2",
+        salt: encoder.encode(this.salt),
+        iterations: 100000,
+        hash: "SHA-256",
+      },
+      initialKey,
+      { name: "AES-GCM", length: 256 },
+      false,
+      ["encrypt", "decrypt"]
+    );
+  }
+
+  /**
+   * Encrypts the provided plaintext string.
+   *
+   * @param plaintext The plaintext string to encrypt.
+   * @returns A base64-encoded string containing the initialization vector and the ciphertext.
+   */
+  public async encrypt(plaintext: string): Promise<string> {
+    const encoder = new TextEncoder();
+
+    // Each AES-GCM encryption requires a unique initialization vector, or
+    // iv. The iv should be 12 bytes long and is safe to store in plaintext
+    // alongside the ciphertext.
+    const iv = crypto.getRandomValues(new Uint8Array(12));
+
+    const ciphertext: ArrayBuffer = await crypto.subtle.encrypt(
+      {
+        name: "AES-GCM",
+        iv,
+      },
+      this.key,
+      encoder.encode(plaintext)
+    );
+
+    // Prepend the iv to the ciphertext, so that it can be retrieved later.
+    return self.btoa(
+      [...new Uint8Array(iv), ...new Uint8Array(ciphertext)]
+        .map((b) => String.fromCharCode(b))
+        .join("")
```

Comment:
Trying to understand this bit ... is it concating the iv and the encrypted data, and then turning it into a string? Maybe worth a chat since you were also talking about how we'll get a blob in the request response (?) while here we need a string ...